### DISCUSSION TITLE
Infiniband interface using 48bit mac instead of 64bit guid

### DISCUSSION THREAD

**Discussion by tdunn3-uco** [ipxe/ipxe#510] [Q&A]
Hello,
I currently have a warewulf cluster which boots over a management network with our infiniband network being used for slurm jobs. I am trying to eliminate the management network by booting over infiniband. I have it working except warewulf uses 64bit guid for infiniband devices and configures dhcp to do the same. The problem is my Qlogic card (qib7322 driver) uses the 48 bit mac so it can't configure itself using dhcp.

Does anyone have suggestions as to how I could tell IPXE to use guid instead of mac?

**Comment by mcb30** [comment]
iPXE handles Infiniband IPoIB devices using the eIPoIB abstraction, under which the devices pretend to have an Ethernet-compatible 48-bit MAC address and packets are rewritten as needed at the link layer.  This is required in order to allow the devices to be used via the UNDI API, which is basically restricted to handling Ethernet devices.

The chaddr field will therefore contain the 48-bit Ethernet-compatible eIPoIB MAC address.  You should find that some other field within the DHCP request does contain the original Infiniband GUID.  How does warewulf configure the DHCP server to recognise the Infiniband devices?

**Reply by tdunn3-uco** [reply]
Thank you for replying. It creates a little host block for each node with lines like "hardware ethernet 00:11:75:00:00:6f:3b:f0;". I can manually edit the generated dhcp file and change the address to 48bit and ipxe will configure itself or I can change the node hwaddr in warewulf and it will go through the whole process of pulling the image and most of warewulf's bootstrap script but it will fail when it tries to lookup the node in warewulf's database because it is designed to detect infiniband devices and use guid for referencing them.

**Reply by mcb30** [reply]
The `hardware ethernet 00:11:75:00:00:6f:3b:f0;` declaration is invalid since the address length is incorrect for an Ethernet link-layer.

I don't have easy access to an Infiniband test setup for the next few days, but you could try using wireshark to examine the contents of iPXE's DHCP request.  There should be a DHCP option that contains the original GUID (or possibly the GID), which you could use to match instead of the `hardware ...` declaration (which matches the DHCP chaddr field).

**Reply by tdunn3-uco** [reply]
Okay, I'll play around with wireshark and see if I can learn something. I guess my next stop after that will be the warewulf forms to learn how to modify how the dhcp file is generated. I'll reply here if I discover a solution. Thanks.

**Reply by tdunn3-uco** [reply]
I was able to look at the packet ipxe is sending out:
```
15:15:25.063774 IP (tos 0x0, ttl 64, id 1024, offset 0, flags [none], proto UDP (17), length 428) 
   0.0.0.0.bootpc > 255.255.255.255.bootps: [udp sum ok] BOOTP/DHCP, Request from 00:11:75:6f:3b:f0 (oui Unknown), length 400, xid 0xd0800c18, secs 16, Flags [Bro
adcast] (0x8000) 
     Client-Ethernet-Address 00:11:75:6f:3b:f0 (oui Unknown) 
     Vendor-rfc1048 Extensions 
       Magic Cookie 0x63825363 
       DHCP-Message Option 53, length 1: Discover 
       MSZ Option 57, length 2: 1472 
       ARCH Option 93, length 2: 0 
       NDI Option 94, length 3: 1.2.1 
       Vendor-Class Option 60, length 32: "PXEClient:Arch:00000:UNDI:002001" 
       User-Class Option 77, length 4:  
         instance#1: ERROR: invalid option 
       Parameter-Request Option 55, length 23:  
         Subnet-Mask, Default-Gateway, Domain-Name-Server, LOG 
         Hostname, Domain-Name, RP, MTU 
         Vendor-Option, Vendor-Class, TFTP, BF 
         Option 119, Option 128, Option 129, Option 130 
         Option 131, Option 132, Option 133, Option 134 
         Option 135, Option 175, Option 203 
       T175 Option 175, length 48: 2969895184,2004034283,50402561,385941794,16847617,17891585,654377241,16846849,35717377,352387352,16849665,17957121 
       Client-ID Option 61, length 7: ether 00:11:75:6f:3b:f0 
       GUID Option 97, length 17: 0.27.162.11.159.143.95.248.10.139.217.28.183.44.16.63.49 
       END Option 255, length 0
```
The problem is, it looks wrong. The GUID is as long as I would expect the Client-ID to be and after I converted it to hex I noticed that the last 6 octets match the mac address for the normal ethernet card on my machine. I assume I have something configured wrong.

**Reply by tdunn3-uco** [reply]
I've continued my testing. Here is the output of the ibstat command on my node:
```
CA 'qib0'
        CA type: InfiniPath_QLE7340
        Number of ports: 1
        Firmware version: 
        Hardware version: 2
        Node GUID: 0x00117500006f3bf0
        System image GUID: 0x00117500006f3bf0
        Port 1:
                State: Active
                Physical state: LinkUp
                Rate: 40
                Base lid: 40
                LMC: 0
                SM lid: 1
                Capability mask: 0x07690868
                Port GUID: 0x00117500006f3bf0
                Link layer: InfiniBand
```
The guid is correct. I'm not sure why that other value is being pulled by ipxe but I now know the correct guid is in there somewhere. Does anyone have any ideas why the wrong guid is being pulled?
